lib/rules/no-hooks-from-ancestor-modules.js

```diff
@@ -59,11 +59,15 @@ module.exports = {
                 isInModuleCallbackBody(node);
         }
 
-        function isFunctionWithIdentifierArg(node) {
-            return node &&
-                node.type === "FunctionExpression" &&
-                node.params.length > 0 &&
-                node.params[0].type === "Identifier";
+        function getNonThisIdentifierFromParams(params) {
+            // In TypeScript, `this` can be passed as the first function parameter to add a type to it,
+            // and we want to ignore that parameter since we're looking for the `hooks` variable.
+            if (params.length === 1 && params[0].type === "Identifier" && params[0].name !== "this") {
```

Comment:
Wouldn't it be simpler to do something like the following:

```js
return params.find(p => p.type === "Identifier" && p.name !== "this");
```

Reply:
I had that before, but we actually only want to check specific argument positions.

Reply:
Can you provide an example of where my suggested logic would return incorrect output, compared with the current logic?

Reply:
If we just use `find` then we could end up returning a node even when we have unexpected arguments present:
```
module('foo', function(123, assert) {});
module('foo', function(this: LocalTestContext, 123, assert) {});
```
Arguably, it's better to fail fast and just return `null` if any unexpected arguments are present, which is why my solution only checks for the assert argument as the first argument or directly after the `this` argument.

Reply:
I'm confused. Is it legal to have literals (rather than identifiers) in function expressions? (Note that these are not function *calls*, where any expression would be legal.)

My working assumption right now is that only identifiers are allowed (I suppose assignment patterns are allowed too, but I don't think QUnit really supports those). So, you'll always one of basically six cases:

1. `module('foo', function() {});
1. `module('foo', function(hooks) {});`
1. `module('foo', function(foo, notActuallyHooks) {});`
1. `module('foo', function(this: LocalTestContext) {});
1. `module('foo', function(this: LocalTestContext, hooks) {});`
1. `module('foo', function(this: LocalTestContext, foo, notActuallyHooks) {});`

That is, you have a matrix of: Is there a `this` parameter present; and then, is there a parameter after that.

QUnit will pass `hooks` in as the first parameter. In the case of Typescript, my understanding is that `this: LocalTestContext` is more or less syntactic sugar to tell the compiler what the type of `this` should be, but in calling code, the first argument is bound to the first parameter *after* the `this` declaration.

So in the third and sixth examples above, `foo` is the real hooks object, even if you have a later parameter called `hooks`.

I believe the logic I suggested should work for all 6 cases:

1. Returns undefined
1. Returns hooks
1. Returns foo
1. Returns undefined
1. Returns hooks
1. Returns foo

Please let me know if I've made a bad assumption anywhere.

Reply:
Ahh. My bad. The literals thing was a silly mistake which invalidates my example.